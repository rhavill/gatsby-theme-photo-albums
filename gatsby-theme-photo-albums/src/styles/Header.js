/** @jsx jsx */
import { jsx } from 'theme-ui'
import PropTypes from 'prop-types'

const Header = ({children}) => 
  <div sx={{variant: 'layout.header'}}>
    {children}
  </div>

Header.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
}

export default Header