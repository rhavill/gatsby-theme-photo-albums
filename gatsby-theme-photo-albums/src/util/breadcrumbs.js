import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs