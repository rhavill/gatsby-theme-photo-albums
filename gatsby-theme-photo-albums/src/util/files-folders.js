const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}