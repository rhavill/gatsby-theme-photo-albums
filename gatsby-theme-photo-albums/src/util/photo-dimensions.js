export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}