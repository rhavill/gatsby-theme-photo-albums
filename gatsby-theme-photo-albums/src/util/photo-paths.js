const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension, ensureLeadingAndTrailingSlash} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => { 
    let previousUrl = null
    let previousUrlWithPage = null
    return indexedForEach(
      (file, i) => {
        const pageNumber = Math.ceil((i + 1) / photosPerPage)
        photoPaths[file.url] = {
          url: getPhotoPathWithPage(pageNumber, file.url),
          previousUrl: previousUrlWithPage,
          nextUrl: null,
          parentUrl: appendPageNumberToUrl(pageNumber, file.parentUrl),
        }
        if (previousUrl) {
          photoPaths[previousUrl].nextUrl = photoPaths[file.url].url
        }
        previousUrl = file.url
        previousUrlWithPage = photoPaths[file.url].url
      }, 
      groupedFiles[relativeDirectory]
    )
  }, keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `page-${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

const appendPageNumberToUrl = (pageNumber, url) => {
  let path = url ? url : null
  if (path && pageNumber > 1) {
    path = `${ensureLeadingAndTrailingSlash(path)}page-${pageNumber}`
  }
  return path
}

module.exports = {
  getPhotoPathsWithPages, 
  getPhotoPathWithPage,
}