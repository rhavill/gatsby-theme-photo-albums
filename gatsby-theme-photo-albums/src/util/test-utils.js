import mergeDeepRight from 'ramda/src/mergeDeepRight'
import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'emotion-theming'

import theme from '../theme'

const create = (element, options) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{element}</ThemeProvider>, options
  )
}

const customRenderer = mergeDeepRight(renderer, {create})

const Wrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}
Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options })

// re-export everything from React Testing Library
export * from '@testing-library/react'

// override react testing library render method and custom react-test-renderer r
export { customRender as render, customRenderer as renderer }