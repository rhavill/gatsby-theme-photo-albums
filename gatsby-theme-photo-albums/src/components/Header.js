import React from 'react'
import PropTypes from 'prop-types'
import {Link} from 'gatsby'
import {Header as ThemeHeader} from 'theme-ui'
import breadcrumbs from '../util/breadcrumbs'

const Header = ({ path, rootTitle }) => {
  const links = breadcrumbs(path, rootTitle)

  return (
    <ThemeHeader>
      <nav>{
        links.map((crumb, i) => (
          <Link key={i} to={crumb.path}>
            {crumb.title}
          </Link>))
      }
      </nav>
    </ThemeHeader>
  )
}

Header.propTypes = {
  path: PropTypes.string.isRequired,
  rootTitle: PropTypes.string,
}

Header.defaultProps = { 
  rootTitle: 'Home',
}

export default Header