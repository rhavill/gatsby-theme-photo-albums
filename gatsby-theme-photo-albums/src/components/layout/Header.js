import React from 'react'
import PropTypes from 'prop-types'
import HeaderStyles from '../../styles/Header'
import FirstBreadcrumbText from './FirstBreadcrumbText'
import Menu from './Menu'
import breadcrumbs from '../../util/breadcrumbs'

const Header = ({ path }) => {
  const links = breadcrumbs(path)
  if (links.length) {
    links[0].title = <FirstBreadcrumbText />
  }
  return (
    <HeaderStyles>
      <Menu links={links} />
    </HeaderStyles>
  )
}

Header.propTypes = {
  path: PropTypes.string.isRequired,
}

export default Header