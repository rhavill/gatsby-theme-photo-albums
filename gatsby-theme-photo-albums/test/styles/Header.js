import React from 'react'
import {cleanup, renderer} from '../../src/util/test-utils'
import Header from '../../src/styles/Header'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Header styles', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Header><div>hello</div></Header>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
