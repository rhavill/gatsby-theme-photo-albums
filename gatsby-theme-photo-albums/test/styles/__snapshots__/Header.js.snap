// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header styles renders correctly 1`] = `
<div
  className="css-1jhlbdc-Header"
>
  <div>
    hello
  </div>
</div>
`;
